src/System.Linq.Parallel/tests/QueryOperators/ForAllTests.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using Xunit;
+
+namespace Test
+{
+    public class ForAllTests
+    {
+        [Theory]
+        [MemberData("Ranges", (object)(new int[] { 1, 2, 16 }), MemberType = typeof(Sources))]
+        [MemberData("Ranges", (object)(new int[] { 1, 2, 16 }), MemberType = typeof(UnorderedSources))]
+        public static void ForAll(Labelled<ParallelQuery<int>> labelled, int count)
+        {
+            ParallelQuery<int> query = labelled.Item;
+            IntegerRangeSet seen = new IntegerRangeSet(0, count);
+            query.ForAll<int>(x => Assert.True(seen.Add(x)));
```

Comment:
I'm aware it's not threadsafe.  I initially decided that the library was probably well-behaved enough it probably wouldn't be an issue, or that any duplicate entries probably had a good enough chance of mutating something already set.
Although since it's not `volatile` Booleans... hmmmm... 
I was really hoping to avoid the overhead of locking.  Especially because that's going to increase test run time.